crates/ruff_linter/src/rules/flake8_quotes/rules/avoidable_escaped_quote.rs

```diff
@@ -0,0 +1,258 @@
+/// It's preferable to avoid escaped quotes in strings. By changing the
+/// outer quote style, you can avoid escaping inner quotes.
+///
+/// ## Example
+/// ```python
+/// foo = 'bar\'s'
+/// ```
+///
+/// Use instead:
+/// ```python
+/// foo = "bar's"
+/// ```
+#[violation]
+pub struct AvoidableEscapedQuote;
+
+impl AlwaysAutofixableViolation for AvoidableEscapedQuote {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Change outer quotes to avoid escaping inner quotes")
+    }
+
+    fn autofix_title(&self) -> String {
+        "Change outer quotes to avoid escaping inner quotes".to_string()
+    }
+}
+
+struct FStringContext {
+    /// Whether to check for escaped quotes in the f-string.
+    check_for_escaped_quote: bool,
+    /// The range of the f-string start token.
+    fstring_start_range: TextRange,
+    /// The ranges of the f-string middle tokens containing escaped quotes.
+    fstring_middle_ranges: Vec<TextRange>,
+}
+
+impl FStringContext {
+    fn new(check_for_escaped_quote: bool, fstring_start_range: TextRange) -> Self {
+        Self {
+            check_for_escaped_quote,
+            fstring_start_range,
+            fstring_middle_ranges: vec![],
+        }
+    }
+
+    /// Update the context to not check for escaped quotes, and clear any
+    /// existing reported ranges.
+    fn do_not_check_for_escaped_quote(&mut self) {
+        self.check_for_escaped_quote = false;
+        self.fstring_middle_ranges.clear();
+    }
+
+    fn push_fstring_middle_range(&mut self, range: TextRange) {
+        self.fstring_middle_ranges.push(range);
+    }
+}
+
+/// Q003
+pub(crate) fn avoidable_escaped_quote(
+    diagnostics: &mut Vec<Diagnostic>,
+    lxr: &[LexResult],
+    locator: &Locator,
+    settings: &LinterSettings,
+) {
+    let quotes_settings = &settings.flake8_quotes;
+    let supports_pep701 = settings.target_version.supports_pep701();
+    let mut fstrings: Vec<FStringContext> = Vec::new();
+    let mut state_machine = StateMachine::default();
+
+    for &(ref tok, tok_range) in lxr.iter().flatten() {
+        let is_docstring = state_machine.consume(tok);
+        if is_docstring {
+            continue;
+        }
+
+        if !supports_pep701 {
+            // If this is a string or a start of a f-string which is inside another
+            // f-string, we won't check for escaped quotes for the entire f-string
+            // if the target version doesn't support PEP 701. For example:
+            //
+            // ```python
+            // f"\"foo\" {'nested'}"
+            // #          ^^^^^^^^
+            // #          We're here
+            // ```
+            //
+            // If we try to fix the above example, the outer and inner quote
+            // will be the same which is invalid pre 3.12:
+            //
+            // ```python
+            // f'"foo" {'nested'}"
+            // ```
+            if matches!(tok, Tok::String { .. } | Tok::FStringStart) {
```

Comment:
In case this is a nested string or f-string: We disable the tracking for the entire outer fstring. Isn't this too late in case we already flagged an escape early in the fstring?

Reply:
Yes, that's why the `do_not_check_for_escaped_quote` method exists to stop checking any further and remove any existing ranges. Note that we don't extend the diagnostics until we reach the end of the f-string.